src/deviceCommunication/commUSB.cpp

```diff
@@ -52,22 +52,21 @@ void CommUSB::sendData(const QByteArray& rawData) {
 };
 
 void CommUSB::readData() {
-    /// @todo add proper parser
-    QByteArray data = serialPort.readAll();
-    COMbuffer += data;
-    QStringList COMBufferList = COMbuffer.split('\n');
-    for (int i = 0; i < COMBufferList.length(); ++i) {
-        QString currentMsg = COMBufferList[i];
-        if (currentMsg.endsWith('\r')) {
-            float force = currentMsg.mid(1, 6).remove('-').toFloat();
-
-            COMbuffer.remove(currentMsg);
-            emit newForceDevice(currTime, force);
-            
-            currTime += 1.0f/40.0f; // TODO: use actual sampling rate
+    COMbuffer += serialPort.readAll();
+    int msgLen = Parser::PACKET_EXPECTED_LEN;
+    while (COMbuffer.length() >= msgLen) {
+        if (COMbuffer[msgLen - 1] == '\r') {
+            extractedMessage = COMbuffer.mid(0, msgLen);
+            bool success = parser.parsePackage(extractedMessage, receivedData);
+            if (success) {
+                emit newSampleDevice(receivedData);
+            }
+            COMbuffer.remove(0, msgLen); // remove parsed package
+        } else {
+            COMbuffer.remove(0, 1);  // remove first byte and try again
         }
     }
```

Comment:
To avoid parsing bytes that will never be a complete Sample, if `parsePackage()` fails:
1. Log a warning that an invalid packet was received.
2. Discard all bytes until the first `\n` in the buffer.
3. Wait for a new `\n`.
4. Try to parse the last `PACKET_EXPECTED_LEN` bytes.
5. If parsing fails, goto 1, else goto 3.

This would only bring a performance improvement on a bad connection (many invalid packets), but logging warnings of invalid packets would be much easier.

Reply:
Additionally, make sure that `COMbuffer` does not grow infinitely. Can be done for example by
1. Wait until `3*PACKET_EXPECTED_LEN` are in the buffer.
2. Discard the first `PACKET_EXPECTED_LEN` bytes.